File: Solutions/51-n-queens/n-queens.cs
public class Solution {
    // Helper function to check if placing a queen at position (row,col) is safe
    private bool IsSafePlace(int n, List<string> nQueens, int row, int col) {
        // Check if there's any queen in the same column above current position
        for (int i = 0; i < n; i++) {
            if (nQueens[i][col] == 'Q') {
                return false;
            }
        }

        // Check upper-left diagonal for any queen
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (nQueens[i][j] == 'Q') {
                return false;
            }
        }

        // Check upper-right diagonal for any queen
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (nQueens[i][j] == 'Q') {
                return false;
            }
        }

        // If no conflicts found, position is safe
        return true;
    }

    // Recursive helper function to solve N-Queens problem
    private void SolveNQueens(int n, IList<IList<string>> output, List<string> nQueens, int row) {
        // Base case: if we've placed queens in all rows, we found a valid solution
        if (row == n) {
            output.Add(new List<string>(nQueens));
            return;
        }

        // Try placing queen in each column of current row
        for (int col = 0; col < n; col++) {
            // If current position is safe
            if (IsSafePlace(n, nQueens, row, col)) {
                // Place queen
                char[] rowChars = nQueens[row].ToCharArray();
                rowChars[col] = 'Q';
                nQueens[row] = new string(rowChars);
                
                // Recursively solve for next row
                SolveNQueens(n, output, nQueens, row + 1);
                
                // Backtrack: remove queen for trying next position
                rowChars[col] = '.';
                nQueens[row] = new string(rowChars);
            }
        }
    }

    // Main function to solve N-Queens problem
    public IList<IList<string>> SolveNQueens(int n) {
        IList<IList<string>> output = new List<IList<string>>();  // Stores all valid solutions
        List<string> nQueens = new List<string>(); // Initialize empty board
        
        // Initialize the board with empty squares
        for (int i = 0; i < n; i++) {
            nQueens.Add(new string('.', n));
        }
        
        SolveNQueens(n, output, nQueens, 0); // Start solving from row 0
        return output;
    }
}

File: Solutions/51-n-queens/n-queens.swift
class Solution {
    // Helper function to check if placing a queen at position (row,col) is safe
    private func isSafePlace(_ n: Int, _ nQueens: [String], _ row: Int, _ col: Int) -> Bool {
        // Check if there's any queen in the same column above current position
        for i in 0..<n {
            if Array(nQueens[i])[col] == "Q" {
                return false
            }
        }
        
        // Check upper-left diagonal for any queen
        var i = row - 1
        var j = col - 1
        while i >= 0 && j >= 0 {
            if Array(nQueens[i])[j] == "Q" {
                return false
            }
            i -= 1
            j -= 1
        }
        
        // Check upper-right diagonal for any queen
        i = row - 1
        j = col + 1
        while i >= 0 && j < n {
            if Array(nQueens[i])[j] == "Q" {
                return false
            }
            i -= 1
            j += 1
        }
        
        // If no conflicts found, position is safe
        return true
    }
    
    // Recursive helper function to solve N-Queens problem
    private func solveNQueens(_ n: Int, _ output: inout [[String]], _ nQueens: inout [String], _ row: Int) {
        // Base case: if we've placed queens in all rows, we found a valid solution
        if row == n {
            output.append(nQueens)
            return
        }
        
        // Try placing queen in each column of current row
        for col in 0..<n {
            // If current position is safe
            if isSafePlace(n, nQueens, row, col) {
                // Place queen
                var rowArray = Array(nQueens[row])
                rowArray[col] = "Q"
                nQueens[row] = String(rowArray)
                
                // Recursively solve for next row
                solveNQueens(n, &output, &nQueens, row + 1)
                
                // Backtrack: remove queen for trying next position
                rowArray[col] = "."
                nQueens[row] = String(rowArray)
            }
        }
    }
    
    // Main function to solve N-Queens problem
    func solveNQueens(_ n: Int) -> [[String]] {
        var output: [[String]] = []  // Stores all valid solutions
        var nQueens = Array(repeating: String(repeating: ".", count: n), count: n) // Initialize empty board
        solveNQueens(n, &output, &nQueens, 0) // Start solving from row 0
        return output
    }
}